### ISSUE TITLE
model.train() and model.eval() gives different results

### ISSUE THREAD

**Issue by pxliang** [facebookresearch/detr#498]
I trained a model and did inference, the model predicts the same bounding boxes for all images with model.eval(). However, the model gives different results with model.train(). Why does the model perform so strange with model.train() and model.eval()? Can someone helps me?

**Comment by ywyue** [comment]
Hi I encounter the same problem, have you solved it?

**Comment by mariiak2021** [comment]
Hi @pxliang and @ywyue I experience the same problem. Did you find any information or solution about it?

**Comment by ywyue** [comment]
Hi, currently during inference, I just comment the line model.eval(). Please let me know if you have better solutions.

**Comment by pxliang** [comment]
I didn't have better solution... I just comment the model.eval() too.

**Comment by mariiak2021** [comment]
I think it's not correct to use without .eval actually... it' not making the fair prediction then.
If I find something - will let you know! Thank you for the fast answer.

**Comment by ywyue** [comment]
Hi @mariiak2021 Why do you think it is not a fair prediciton?

**Comment by mariiak2021** [comment]
> Hi @mariiak2021 Why do you think it is not a fair prediciton?

Because .eval should be used during inference.

**Comment by qqaatw** [comment]
Spontaneously seeing this issue. The results differ because operators like `dropout` behave different between train and eval. When eval is set, dropout is disabled in order to provide deterministic results. see https://pytorch.org/docs/stable/generated/torch.nn.Dropout.html

**Comment by mariiak2021** [comment]
Hi @qqaatw yes - for sure the dropout layers behave differently in eval and train modes. This is not the issue.
The problem is that the model is predicting nearly same outputs for any input in evaluation mode, if ANY dropout value was used during training.

**Comment by lohmander** [comment]
I'm experiencing a similar issue, setting all the dropout layers to train-mode (as discussed [here](https://discuss.pytorch.org/t/using-dropout-in-evaluation-mode/27721)) yields results in line with training mode, however I'm not sure this is an appropriate approach...

```python
model.eval()
for m in model.modules():
  if m.__class__.__name__.startswith('Dropout'):
    m.train()
```

**Comment by mariiak2021** [comment]
Hi @lohmander thank you for sharing! But I think it's not appropriate. What you are doing is basically using dropout in evaluation mode as well, which should be avoided.
I try last days to train the model with dropout = 0.0 and the results are surprisingly good. The model behaves correct during .eval() mode as well. In my case training without dropout produces good enough results and not overfitting. 
I will investigate further why it could happen anyway.

**Comment by lohmander** [comment]
Hi @mariiak2021, thanks for the tip about training without dropout. My snippet was more of a dirty trick that seemed to at least confirm that it's the dropout that's causing the issue. However, from what I can tell they do indeed use it for the pretrained weights, and it works fine in eval mode (so without dropout applied). Do you have any theories to investigate, because this is a pressing issue for us as well?

**Comment by mariiak2021** [comment]
Hi @lohmander, logically the dropout used during training should NOT be an issue. As we have such a problem, either there is a problem with feature extraction (so that model is not learning anything actually) or some code bugs I guess.
I found information, that nearly same outputs are normal for the training beginning (in .eval() mode) and they will change with the training progress. So did you train your model long enough?

**Comment by mostafaelhoushi** [comment]
I have notiecd a similar issue as well (on another model though). However, if you train for more epochs, the output during evaluation will converge to the output of training.

This is Dropout by design. The whole idea behind dropout is that you won’t get the exact output for the same input during training and evaluation. In the first epochs of training, the difference will be large, but in later epochs, the model is trained to be resilient such that when different weights in its model are turned off, the output won’t change much.

**Comment by mariiak2021** [comment]
Hi Mostafa, For how long epochs did you train the model?In my case the eval and train outputs don't converge even after huge amount of epochs (1000).Please let me know. Best,Mariia 26.02.2023, 15:52, "Mostafa Elhoushi" ***@***.***>: I have notiecd a similar issue as well (on another model though). However, if you train for more epochs, the output during evaluation will converge to the output of training.This is Dropout by design. The whole idea behind dropout is that you won’t get the exact output for the same input during training and evaluation. In the first epochs of training, the difference will be large, but in later epochs, the model is trained to be resilient such that when different weights in its model are turned off, the output won’t change much.—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

**Comment by mostafaelhoushi** [comment]
Hi @mariiak2021 . Actually, I was facing the problem on a totally different training task. I found this GitHub issue when Googling the issue. 
I actaully removed my comment because it is not related to the detr repo.

Maybe consider taking the first dropout layer in the model and log its input and output in both training and eval to analyze what's going on?

**Comment by nairbv** [comment]
@pxliang you might want to change the title of the issue. The title itself looks like a description of expected behavior due to dropout, but all inference calls giving the same result isn't.

**Comment by zx970723** [comment]
我在自己的数据集中遇到了相同的问题：
依次尝试过：
1、batch size调大
2、数据归一化、标准化
3、去掉backbone中BN
4、dropout=0
在完成3、4后得到符合预期的效果
但是这样并不利于模型训练，让人很迷茫
而且我将coco val数据集同时用于训练和预测，问题表现一致，排除了数据分布不同导致的问题

**Comment by tldrafael** [comment]
I'm also facing this issue. Although, I was implementing a small and educative version of DETR from scratch and using the `nn.Transformer` module.
